### DISCUSSION TITLE
Do you need (should you have) every version of .NET installed?

### DISCUSSION THREAD

**Discussion by Sin-Shadow-Fox** [dotnet/runtime#103376] [Q&A] [answered]
Apologies if this is in the wrong section but there does not appear to be an option to open a new discussion in this repository.

I was having a discussion with [EatonZ](https://github.com/EatonZ) about which version of .NET i needed when [hamarb123](https://github.com/hamarb123) said 

> Each release version of .NET can be installed side-by-side, and furthermore, if an app is built for .NET 7, it won't run unless you have 7 installed (usually), even if you have 8 installed.

So should i have every version of .NET installed? And if so, does anyone know where i can get the desktop runtime for 2.2, 2.1, 2.0, 1.1 and 1.0?

**Comment by julealgon** [comment]
> Apologies if this is in the wrong section but there does not appear to be an option to open a new discussion in this repository.

Not sure what you mean. It is available there:
![image](https://github.com/dotnet/runtime/assets/461579/6499cb42-740e-4280-9e27-6b8a9822dcaa)
https://github.com/dotnet/runtime/discussions

And yes, this would be better as a question there than an issue.

**Comment by Sin-Shadow-Fox** [comment]
> > Apologies if this is in the wrong section but there does not appear to be an option to open a new discussion in this repository.
> 
> Not sure what you mean. It is available there: ![image](https://private-user-images.githubusercontent.com/461579/339116208-6499cb42-740e-4280-9e27-6b8a9822dcaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyMjM0NTcsIm5iZiI6MTcxODIyMzE1NywicGF0aCI6Ii80NjE1NzkvMzM5MTE2MjA4LTY0OTljYjQyLTc0MGUtNDI4MC05ZTI3LTZiOGE5ODIyZGNhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxMlQyMDEyMzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOWYzYzZhZTJjMzhiZmUwMTMwMjA5YjNjZTBlMTY5YWVjMjcyOWJjNTBkOGFiOWM5MWRkYTJjN2Q2ODcxMTY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.KJTmgAlNaqMBiiN10ityaFvfBZzH56fBIcexRyd7s-s) https://github.com/dotnet/runtime/discussions
> 
> And yes, this would be better as a question there than an issue.

![Capture](https://github.com/dotnet/runtime/assets/96899282/d1bebcdf-565d-44be-80a7-9fd2d1aeb766)

**Comment by dotnet-policy-service** [comment]
Tagging subscribers to this area: @dotnet/area-meta
See info in [area-owners.md](https://github.com/dotnet/runtime/blob/main/docs/area-owners.md) if you want to be subscribed.

**Comment by jkotas** [comment]
> So should i have every version of .NET installed? 

No. You only need to install .NET runtime versions used by the apps that you run.

**Comment by jkotas** [comment]
If the app comes with an installer, the installer should take care of this for you.

**Comment by Sin-Shadow-Fox** [comment]
> > So should i have every version of .NET installed?
> 
> No. You only need to install .NET runtime versions used by the apps that you run.

Ok, then how many versions of .NET do apps use?

Also thank you for adding the question label.



> If the app comes with an installer, the installer should take care of this for you.

Before i was provided with links to install .NET from [EatonZ](https://github.com/EatonZ), i didn't even have .NET installed on my computer at all.

**Comment by jkotas** [comment]
> Ok, then how many versions of .NET do apps use?

A given app typically uses at most one specific .NET version.

Well behaving apps should come with an installer that takes care of this for you.

If the app does not come with an installer, the instructions for manual installation of the app should like the .NET version that you have to install manually. This is a question for the app publisher.

**Comment by julealgon** [comment]
> > > Apologies if this is in the wrong section but there does not appear to be an option to open a new discussion in this repository.
> > 
> > 
> > Not sure what you mean. It is available there: ![image](https://private-user-images.githubusercontent.com/461579/339116208-6499cb42-740e-4280-9e27-6b8a9822dcaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyMjM0NTcsIm5iZiI6MTcxODIyMzE1NywicGF0aCI6Ii80NjE1NzkvMzM5MTE2MjA4LTY0OTljYjQyLTc0MGUtNDI4MC05ZTI3LTZiOGE5ODIyZGNhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxMlQyMDEyMzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOWYzYzZhZTJjMzhiZmUwMTMwMjA5YjNjZTBlMTY5YWVjMjcyOWJjNTBkOGFiOWM5MWRkYTJjN2Q2ODcxMTY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.KJTmgAlNaqMBiiN10ityaFvfBZzH56fBIcexRyd7s-s) https://github.com/dotnet/runtime/discussions
> > And yes, this would be better as a question there than an issue.
> 
> ![Capture](https://private-user-images.githubusercontent.com/96899282/339119052-d1bebcdf-565d-44be-80a7-9fd2d1aeb766.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyMjUwNzUsIm5iZiI6MTcxODIyNDc3NSwicGF0aCI6Ii85Njg5OTI4Mi8zMzkxMTkwNTItZDFiZWJjZGYtNTY1ZC00NGJlLTgwYTctOWZkMmQxYWViNzY2LlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDIwMzkzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyZTJjM2VlYWMxZjc1NTcwN2MwNWNkNDM5NTVlM2Y2NDhmYWY2ZGYzNzViODMwNTdjNmFjMTg2YTlhZjY5NjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6oPq-6gDxC-x-YLdM_HM-fwpPDLPUMAChQXVKfDk7do)

That's not the `runtime` repo @Sin-Shadow-Fox , it's the `core` one. The "Discussions" feature is configured on a per-repo basis.

**Comment by Sin-Shadow-Fox** [comment]
> > Ok, then how many versions of .NET do apps use?
> 
> A given app typically uses at most one specific .NET version.

Well I'm not asking about one specific app. I'm asking about all of them.

> Well behaving apps should come with an installer that takes care of this for you.
> 
> If the app does not come with an installer, the instructions for manual installation of the app should like the .NET version that you have to install manually. This is a question for the app publisher.

Yeah I'm still waiting on a response from them > https://github.com/SteamRE/DepotDownloader/discussions/505



> > > > Apologies if this is in the wrong section but there does not appear to be an option to open a new discussion in this repository.
> > > 
> > > 
> > > Not sure what you mean. It is available there: ![image](https://private-user-images.githubusercontent.com/461579/339116208-6499cb42-740e-4280-9e27-6b8a9822dcaa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyMjM0NTcsIm5iZiI6MTcxODIyMzE1NywicGF0aCI6Ii80NjE1NzkvMzM5MTE2MjA4LTY0OTljYjQyLTc0MGUtNDI4MC05ZTI3LTZiOGE5ODIyZGNhYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNjEyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDYxMlQyMDEyMzdaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mOWYzYzZhZTJjMzhiZmUwMTMwMjA5YjNjZTBlMTY5YWVjMjcyOWJjNTBkOGFiOWM5MWRkYTJjN2Q2ODcxMTY2JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.KJTmgAlNaqMBiiN10ityaFvfBZzH56fBIcexRyd7s-s) https://github.com/dotnet/runtime/discussions
> > > And yes, this would be better as a question there than an issue.
> > 
> > 
> > ![Capture](https://private-user-images.githubusercontent.com/96899282/339119052-d1bebcdf-565d-44be-80a7-9fd2d1aeb766.PNG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTgyMjUwNzUsIm5iZiI6MTcxODIyNDc3NSwicGF0aCI6Ii85Njg5OTI4Mi8zMzkxMTkwNTItZDFiZWJjZGYtNTY1ZC00NGJlLTgwYTctOWZkMmQxYWViNzY2LlBORz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA2MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNjEyVDIwMzkzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMyZTJjM2VlYWMxZjc1NTcwN2MwNWNkNDM5NTVlM2Y2NDhmYWY2ZGYzNzViODMwNTdjNmFjMTg2YTlhZjY5NjkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6oPq-6gDxC-x-YLdM_HM-fwpPDLPUMAChQXVKfDk7do)
> 
> That's not the `runtime` repo @Sin-Shadow-Fox , it's the `core` one. The "Discussions" feature is configured on a per-repo basis.

Yeah i have no idea what any of that means. All i know is i was having a convo with someone here and then that got locked so i opened an issue to ask my question.

**Reply by julealgon** [reply]
You should put a little bit of effort into understanding it then before opening issues. It's a bit disrespectful/dismissive on your part to apologize upfront but not put any effort into trying to understand it first.

**Reply by Sin-Shadow-Fox** [reply]
> You should put a little bit of effort into understanding it then before opening issues. It's a bit disrespectful/dismissive on your part to apologize upfront but not put any effort into trying to understand it first.

huh? Sorry i don't know any programming. Maybe try asking someone else?

**Comment by elgonzo** [comment, marked as answer]
> Ok, then how many versions of .NET do apps use?

> Well I'm not asking about one specific app. I'm asking about all of them.

As there is for each released .NET version at least one app somewhere in the world using that .NET version, the only logical and true answer is then: All ever released .NET versions.

**Reply by Sin-Shadow-Fox** [reply]
Yeah, that's what i figured, so i just wanted to ask to make sure. Now the only problem is getting the installs for 2.2 and backwards.

**Reply by jkotas** [reply]
The installers for all .NET runtime versions are available at https://dotnet.microsoft.com/en-us/download/dotnet . You have to click on "Out of support versions" to see them. 

It is not particularly good idea to install out-of-support .NET versions. They may expose you to security issues or not work at all on current OSes.

**Reply by Sin-Shadow-Fox** [reply]
Actually that site only has the desktop runtimes for 3.0 and above.

Also what would you recommend instead of installing older versions?

**Reply by elgonzo** [reply]
> Actually that site only has the desktop runtimes for 3.0 and above.

Okay. Then how do you explain this screenshot from https://dotnet.microsoft.com/en-us/download/dotnet?
![NetDownloads](https://github.com/dotnet/runtime/assets/6129198/821a825e-9135-4c75-8139-8bc183af506c)

**Reply by Sin-Shadow-Fox** [reply]
Easy, that's the version page. If you go into each of those you'll see the desktop runtime for each version for 3.0 up to 8.0. But if you go into any of the ones before 3.0 you'll see that the desktop runtime is missing.

**Reply by elgonzo** [reply]
Oops, didn't realize that you looked for desktop runtimes specifically. There are no desktop runtimes available for download for .NET Core 2.x and below because there were simply no desktop runtimes for those .NET Core versions released. Desktop runtimes were only introduced with .NET Core 3.0 (see here for a relevant MS blog article about this: https://devblogs.microsoft.com/dotnet/net-core-3-and-support-for-windows-desktop-applications/). In other words, any programs developed with .NET Core 2.x and below were console apps, web backends, services or some such, but no desktop GUI apps.

Also, various .NET Framework versions can be found here: https://dotnet.microsoft.com/en-us/download/dotnet-framework
(The runtime for .NET Framework 3.5 should also include the runtimes for the older .NET Framework 3.0 and 2.0 if my memory serves me right. Also, pay attention to the note on the .NET Framework 3.5 download page regarding its installation on Windows 10/11.)

**Reply by Sin-Shadow-Fox** [reply]
Oh i thought i needed desktop runtimes specifically because that is what i was [provided with](https://github.com/dotnet/core/issues/7556) by [EatonZ](https://github.com/EatonZ). Also i have no idea what that article is talking about. most of this stuff goes over my head.

Also what is ,NET framework? Is it important?

**Reply by elgonzo** [reply]
> what is ,NET framework

It's the brand/product name of earlier .NET versions that were Windows-exclusive. Parallel to later .NET Framework 4.x versions, there was .NET Core introduced, which was cross-platform (and with its source code under a permissive license) but with a limited feature set compared to .NET Framework. For a while, .NET Framework 4.x versions and .NET Core versions existed and were released in parallel. Both .NET Framework and .NET Core were then eventually discontinued, with the new product philosophy becoming a single full-featured, cross-platform and open-source product which would simply be named .NET, the first version of which was .NET 5..

<br>

> Oh i thought i needed desktop runtimes specifically because that is what i was https://github.com/dotnet/core/issues/7556 by [EatonZ](https://github.com/EatonZ).

I have no idea what ghosts you are trying to chase. Do you remember what you asked here?
```
Ok, then how many versions of .NET do apps use?
```
```
Well I'm not asking about one specific app. I'm asking about all of them.
```
Nothing of what you asked here has anything to do with the advice EatonZ provided in the issue thread you linked to. 
Of course it's your prerogative to install any and every .NET version you can get your hands on for whatever reasons you fancy, it's entirely up to you what you want.

But to be bluntly honest, if you ever want to reach success in what you appear to be trying to solve i don't think its helpful if you are ignoring the advice given by jkotas here in this thread:

> You only need to install .NET runtime versions used by the apps that you run.
> Well behaving apps should come with an installer that takes care of this for you.
> If the app does not come with an installer, the instructions for manual installation of the app should like the .NET version that you have to install manually. This is a question for the app publisher.

or ignoring the advice given by agocke in the issue thread you linked to:
> you should ask your questions to the developers of the DepotDownloader tool https://github.com/SteamRE/DepotDownloader

**Reply by Sin-Shadow-Fox** [reply]
> > what is ,NET framework
> 
> It's the brand/product name of earlier .NET versions that were Windows-exclusive. Parallel to later .NET Framework 4.x versions, there was .NET Core introduced, which was cross-platform (and with its source code under a permissive license) but with a limited feature set compared to .NET Framework. For a while, .NET Framework 4.x versions and .NET Core versions existed and were released in parallel. Both .NET Framework and .NET Core were then eventually discontinued, with the new product philosophy becoming a single full-featured, cross-platform and open-source product which would simply be named .NET, the first version of which was .NET 5..

I . . . have no idea what any of that means.

> > Oh i thought i needed desktop runtimes specifically because that is what i was [dotnet/core#7556](https://github.com/dotnet/core/issues/7556) by [EatonZ](https://github.com/EatonZ).
> 
> I have no idea what ghosts you are trying to chase. Do you remember what you asked here?

I'm not familiar with this phrase. All I'm trying to do is to make sure .NET works. Not sure what that has to do with chasing ghosts.

> 
> ```
> Ok, then how many versions of .NET do apps use?
> ```
> 
> ```
> Well I'm not asking about one specific app. I'm asking about all of them.
> ```
> 
> Nothing of what you asked here has anything to do with the advice EatonZ provided in the issue thread you linked to. 

No, most of this is in relation to what [hamarb123](https://github.com/hamarb123) [said](https://github.com/dotnet/core/issues/7556#issuecomment-2161791979).

> Of course it's your prerogative to install any and every .NET version you can get your hands on for whatever reasons you fancy, it's entirely up to you what you want.

But what i want is entirely dependent on what i need. And that is something i don't know, hence why I'm here asking questions and doing my best to learn.

> But to be bluntly honest, if you ever want to reach success in what you appear to be trying to solve i don't think its helpful if you are ignoring the advice given by jkotas here in this thread:
> 
> > You only need to install .NET runtime versions used by the apps that you run.
> > Well behaving apps should come with an installer that takes care of this for you.
> > If the app does not come with an installer, the instructions for manual installation of the app should like the .NET version that you have to install manually. This is a question for the app publisher.

I did follow this advice > https://github.com/SteamRE/DepotDownloader/discussions/505

> or ignoring the advice given by agocke in the issue thread you linked to:
> 
> > you should ask your questions to the developers of the DepotDownloader tool https://github.com/SteamRE/DepotDownloader

Again, i did > https://github.com/SteamRE/DepotDownloader/discussions/505

**Reply by Sin-Shadow-Fox** [reply]
@elgonzo If your still monitoring this discussion, would you please tell me what the ideal versions of .NET 1.0 through 8.0 are if not the desktop runtimes as i believe my discussion is about to get shut down by the collaborators.

**Comment by agocke** [comment]
@Sin-Shadow-Fox, the .NET runtime is a tool for programmers to write programs. As an analogy, you can say that we are screwdriver manufacturers, and people use the screwdrivers to build things. When someone writes a program, they should either screw everything in for you, or bundle the screwdriver needed if some assembly is required.

You are currently asking the screwdriver manufacturer which screwdriver the builder of your chair used. We don't know. We can't possibly know. We will not be able to answer your questions. If you want more information, you will need to go to the author of your program. Only they can help you by either providing you with a copy of the .NET runtime that you need, or by stating exactly what you need to download.

**Reply by Sin-Shadow-Fox** [reply]
This is actually a really good analogy but i would disagree with what I'm asking. This very specific discussion is me asking (using your analogy) How many different types of screwdrivers i would need going forward, not just with my steam issue but with all other issues going forward (i actually have an EZARC mini set of screwdriver heads irl, just in case)

Also, i did ask the author what i need specifically for their program and I'm still waiting on a response > https://github.com/SteamRE/DepotDownloader/discussions/505

**Reply by agocke** [reply]
Based on the `README` file in that repo, it says

> Steam depot downloader utilizing the SteamKit2 library. Supports .NET 8.0

So I would guess that .NET 8 is what the app requires.

**Reply by Sin-Shadow-Fox** [reply]
Well I've got that so either i have the wrong version of 8.0 or i'm missing something, so hopefully the repo owner or one of it's members sees fit to help guide me. Either way, like i said, this discussion is unrelated to that repo or it's project.

**Reply by agocke** [reply]
> This very specific discussion is me asking (using your analogy) How many different types of screwdrivers i would need going forward, not just with my steam issue but with all other issues going forward

Well, then the answer is none (maybe you'll never need a screwdriver), all of them (maybe you'll use every type of screw in existence) or somewhere in the middle. There are thousands of developers tools in existence, many of which cost thousands of dollars a piece. Is it your intention to install all of them, just in case a piece of software you use in the future doesn't appropriately bundle or document its supporting libraries?

**Reply by CyrusNajmabadi** [reply]
> How many different types of screwdrivers i would need going forward, not just with my steam issue but with all other issues going forward (i actually have an EZARC mini set of screwdriver heads irl, just in case)

Since we have no idea which chairs you have now, or will have in the future, we cannot possibly answer how many different types of screwdrivers you would need going Forward.

**Reply by Sin-Shadow-Fox** [reply]
> > This very specific discussion is me asking (using your analogy) How many different types of screwdrivers i would need going forward, not just with my steam issue but with all other issues going forward
> 
> Well, then the answer is none (maybe you'll never need a screwdriver), all of them (maybe you'll use every type of screw in existence) or somewhere in the middle. There are thousands of developers tools in existence, many of which cost thousands of dollars a piece. Is it your intention to install all of them, just in case a piece of software you use in the future doesn't appropriately bundle or document its supporting libraries?

I can only handle one hurdle at a time. Today's hurdle? .NET. Tomorrows hurdle? I'll deal with that then.

**Reply by Sin-Shadow-Fox** [reply]
> > How many different types of screwdrivers i would need going forward, not just with my steam issue but with all other issues going forward (i actually have an EZARC mini set of screwdriver heads irl, just in case)
> 
> Since we have no idea which chairs you have now, or will have in the future, we cannot possibly answer how many different types of screwdrivers you would need going Forward.

As i said, it's not about the chair or chairs i have now. It's about the chairs i could realistically own in the future.

**Reply by agocke** [reply]
You should not pre-emptively install any versions of .NET. If and when you acquire software that requires a specific version, you should install that version.

**Reply by CyrusNajmabadi** [reply]
> It's about the chairs i could realistically own in the future.

You could realistically own all of them. So you might realistically need all screwdrivers.  

What you want to do with that info is up to you. My recommendation is you do what everyone else does and only get the version of .net when you have a need for it. Same as you would for a screwdriver.

**Reply by Sin-Shadow-Fox** [reply]
[elgonzo](https://github.com/dotnet/runtime/discussions/103376#discussioncomment-9756558)
> > Ok, then how many versions of .NET do apps use?
> 
> > Well I'm not asking about one specific app. I'm asking about all of them.
> 
> As there is for each released .NET version at least one app somewhere in the world using that .NET version, the only logical and true answer is then: All ever released .NET versions.
